ext/afform/core/Civi/Afform/PageTokenCredential.php

```diff
@@ -0,0 +1,189 @@
+   *
+   * @param \Civi\Authx\CheckCredentialEvent $check
+   *
+   * @return void
+   */
+  public function afformPageToken(CheckCredentialEvent $check) {
+    if ($check->credFormat === 'Bearer') {
+      try {
+        $claims = \Civi::service('crypto.jwt')->decode($check->credValue);
+        $scopes = isset($claims['scope']) ? explode(' ', $claims['scope']) : [];
+        if (!in_array('afform', $scopes)) {
+          // This is not an afform JWT. Proceed to check any other token sources.
+          return;
+        }
+        if (empty($claims['sub']) || substr($claims['sub'], 0, 4) !== 'cid:') {
+          $check->reject('Malformed JWT. Must specify the contact ID.');
+        }
+        else {
+          $contactId = substr($claims['sub'], 4);
+          if ($this->checkAllowedRoute($check->getRequestPath(), $claims)) {
+            $check->accept(['contactId' => $contactId, 'credType' => 'jwt', 'jwt' => $claims]);
+          }
+          else {
+            $check->reject('JWT specifies a different form or route');
+          }
+        }
+      }
+      catch (CryptoException $e) {
+        // TODO: Is responding that its expired a security risk?
+        if (strpos($e->getMessage(), 'Expired token') !== FALSE) {
+          $check->reject('Expired token');
+        }
+
+        // Not a valid AuthX JWT. Proceed to check any other token sources.
+      }
+    }
+
+  }
+
+  /**
+   * When processing CRM_Core_Invoke, check to see if our token allows us to handle this request.
+   *
+   * @param string $route
+   * @param array $jwt
+   * @return bool
+   * @throws \CRM_Core_Exception
+   * @throws \Civi\API\Exception\UnauthorizedException
+   */
+  public function checkAllowedRoute(string $route, array $jwt): bool {
+    $allowedForm = $jwt['afform'];
+
+    $ajaxRoutes = $this->getAllowedRoutes();
+    foreach ($ajaxRoutes as $regex => $routeInfo) {
+      if (preg_match($regex, $route)) {
+        $parsed = json_decode(\CRM_Utils_Request::retrieve('params', 'String'), 1);
+        if (empty($parsed)) {
+          \Civi::log()->warning("Malformed request. Routes matching $regex must submit params as JSON field.");
+          return FALSE;
+        }
+
+        $extraFields = array_diff(array_keys($parsed), $routeInfo['allowFields']);
+        if (!empty($extraFields)) {
+          \Civi::log()->warning("Malformed request. Routes matching $regex only support these input fields: " . json_encode($routeInfo['allowFields']));
+          return FALSE;
+        }
+
+        $actualForm = $parsed[$routeInfo['nameField']] ?? NULL;
+        if ($actualForm !== $allowedForm) {
+          \Civi::log()->warning("Malformed request. Requested form ($actualForm) does not match allowed name ($allowedForm).");
+          return FALSE;
+        }
+
+        return TRUE;
+      }
+    }
+
+    // Actually, we may not need this? aiming for model where top page-request auth is irrelevant to subrequests...
```

Comment:
I actually don't see at the moment how the token would make it here on the initial request.

It's passed in `_aff` param and the only place that seems to go is to the `onInvoke` above? 

If we want to use this doesn't it need to be picked up in the same way an `_authx` param is in `Authenticator::on_civi_invoke_auth` ?

Reply:
So currently if the form doesn't have anonymous access then I get access denied, even if I have a token.

I could get it working with:
```
diff --git a/ext/authx/Civi/Authx/Authenticator.php b/ext/authx/Civi/Authx/Authenticator.php
index d961b3a8a6..50cd5104bf 100644
--- a/ext/authx/Civi/Authx/Authenticator.php
+++ b/ext/authx/Civi/Authx/Authenticator.php
@@ -63,6 +63,10 @@ class Authenticator extends AutoService implements HookInterface {
         _authx_redact(['_authx']);
       }
     }
+    if (!empty($params['_aff'])) {
+        $this->auth($e, ['flow' => 'param', 'cred' => $params['_aff'], 'siteKey' => $siteKey]);
+        _authx_redact(['_aff']);
+    }
   }
 
   /**
```

Which then meant it got picked up here.

Reply:
I also tried getting my head around the "independent top page request" model.

If I've understood correctly, I create a form where the form itself is anonymously accessible, but it behaves differently for a specific-user.

So I created an anonymous form to create some entities with Role based access, and then tried submitting with/without the `_aff` token. Without token - the entities don't get created as the anonymous user doesn't have permission. With the token, it works!

But I don't see the harm of authenticating the top level request using the token. And it's quite weird to have to remember that you have to make the form world-accessible; and maybe sometimes you don't want it to be, and then how do you provide that indepedent authentication?

In essence I think the Form-Based / User-Based access controls are already pretty hard to get one's head around, so seems like splitting the form page and form api authentication adds another dimension of complexity. So maybe to keep bundled for now?

Reply:
@ufundo Oh, very good point! The admin-experience (configuring perms) will be more forgiving if the token confers access to top-level page.

In fact, it is probably useful to be able to express the distinction between:

* "`civicrm/abc` allows fully anonymous usage; it can be used *with or without tokens*"
* "`civicrm/abc` is not anonymous; but it can be used with tokens" 

And there's no particular drawback. Yes, maybe someday we'd want to take a specific/anonymous form and export a static bundle of JS/HTML. But including auth here doesn't actually prevent that. Going down that path would require more/different QA, but the mechanisms can coexist.

I'll play with your patch to confirm (and hopefully include in the next push).

Reply:
It's still probably worthwhile to make this work, but I think I found trade-off regarding masquerade scenarios -- e.g. if you're logged in as contact 100 and click a link with `?_aff=JWT{...cid=200...}`, then what happens? Where does it draw the line between the session-level-user and the page-level-user?

Intuitively, I think it plays a bit easier if the Afform JS/HTML content is essentially anonymous -- then we only need to think about the AJAX calls. And you can imagine different AJAX calls running with different identities. (*For central page-content, the AJAX calls use the page-token; for peripheral page-content -- eg navbar -- the AJAX calls use your session-token.*) Hard to imagine what you do with the top-level page if needs to simultaneously use permissions of both session-level-user and page-level-user.

Still, in either case, I think masquerade is tricky. If we want something nice, we have to put it back in the oven to bake longer...

---

Of course, "do nothing" is bad option. Right now, it abends and prints 1-line:

```
HTTP 401 Cannot login. Session already active.
```

Maybe in the meantime, a minimal improvement would be to show a regular HTML page with a message like:

```html
<p>You are trying to view a page as another user:</p>

<center><a href="the-full-url-with-auth-token">https://example.com/civicrm/my-form</a></center>

<p>To continue as the other user, please right-click the link and open it in a private window.</p>
```

It's not pretty, but it's more coherent...

Reply:
(*This comment originally misplaced on a different subthread. Re-placing in correct context*)

I added a WIP-commit for the idea of using `_aff` token to identity for the main page-view. (It moved a little bit to better respect module-boundaries, but still basically the same thing.) However, this technique is having an undesirable side-effect -- it causes the user-session to be authenticated. I'm not entirely sure why (and getting too sleep to dig into that). Hence the WIP-commit.

Reply:
> e.g. if you're logged in as contact 100 and click a link with ?_aff=JWT{...cid=200...}, then what happens

As you say, currently you get an error.

To me this seems like a known, pre-existing bug: https://lab.civicrm.org/dev/core/-/issues/4463

I def think it would be good to fix, but it seems outside of the scope of this PR.

Won't it affect the AJAX requests anyway? So a bit orthogonal to the question of which requests are in/out of the tokens authorising power?